DateTimeEditor tests should have conditional skip

Two issues related to the `DateTimeEditor` tests:

- The tests are currently unconditionally skipped. Instead, they should check whether the `DateTimeEditor` is available, and only skip when it's not available.

- After I unskip, if I run against TraitsUI master, I get a test failure:

```
======================================================================
FAIL: test_editor_caching (traits.tests.test_editor_factories.TestDateTimeEditor)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/traits/tests/test_editor_factories.py", line 55, in test_editor_caching
    self.assertIs(editor_1, editor_2)
AssertionError: <traitsui.editors.datetime_editor.DateTimeEditor object at 0x12841d540> is not <traitsui.editors.datetime_editor.DateTimeEditor object at 0x1285f5cc0>
```

@midhun-pm Would you be able to look into this?

--- a/traits/editor_factories.py
+++ b/traits/editor_factories.py
@@ -23,7 +23,6 @@ SourceCodeEditor = None
 HTMLTextEditor = None
 PythonShellEditor = None
 DateEditor = None
-DateTimeEditor = None
 TimeEditor = None
 
 # -------------------------------------------------------------------------------

--- a/traits/tests/test_editor_factories.py
+++ b/traits/tests/test_editor_factories.py
@@ -62,14 +62,6 @@ class TestDateEditor(SimpleEditorTestMixin, unittest.TestCase):
     factory_name = "date_editor"
 
 
-@unittest.skip("DateTimeEditor does not exist in traitsui")
-@requires_traitsui
-class TestDateTimeEditor(SimpleEditorTestMixin, unittest.TestCase):
-    cache_name = "DateTimeEditor"
-    traitsui_name = "DateTimeEditor"
-    factory_name = "datetime_editor"
-
-
 @requires_traitsui
 class TestCodeEditor(SimpleEditorTestMixin, unittest.TestCase):
     cache_name = "SourceCodeEditor"

--- a/traits/tests/test_editor_factories.py
+++ b/traits/tests/test_editor_factories.py
@@ -36,10 +36,6 @@ class SimpleEditorTestMixin:
         self.factory = getattr(traits.editor_factories, self.factory_name)
         self.traitsui_factory = getattr(traitsui.api, self.traitsui_name)
 
-    def tearDown(self):
-        import traits.editor_factories
-        setattr(traits.editor_factories, self.cache_name, None)
-
     def test_editor(self):
         editor = self.factory()
 
@@ -47,6 +43,13 @@ class SimpleEditorTestMixin:
             self.assertIsInstance(editor, traitsui.api.BasicEditorFactory)
         else:
             self.assertIsInstance(editor, self.traitsui_factory)
+
+
+class SimpleEditorWithCachingTestMixin(SimpleEditorTestMixin):
+
+    def tearDown(self):
+        import traits.editor_factories
+        setattr(traits.editor_factories, self.cache_name, None)
 
     def test_editor_caching(self):
         editor_1 = self.factory()
@@ -56,35 +59,43 @@ class SimpleEditorTestMixin:
 
 
 @requires_traitsui
-class TestDateEditor(SimpleEditorTestMixin, unittest.TestCase):
+class TestDateEditor(SimpleEditorWithCachingTestMixin, unittest.TestCase):
     cache_name = "DateEditor"
     traitsui_name = "DateEditor"
     factory_name = "date_editor"
 
 
 @requires_traitsui
-class TestCodeEditor(SimpleEditorTestMixin, unittest.TestCase):
+@unittest.skipIf(getattr(traitsui.api, 'DatetimeEditor', None) is None,
+                 "DatetimeEditor is not in traitsui.api")
+class TestDateTimeEditor(SimpleEditorTestMixin, unittest.TestCase):
+    traitsui_name = "DatetimeEditor"
+    factory_name = "datetime_editor"
+
+
+@requires_traitsui
+class TestCodeEditor(SimpleEditorWithCachingTestMixin, unittest.TestCase):
     cache_name = "SourceCodeEditor"
     traitsui_name = "CodeEditor"
     factory_name = "code_editor"
 
 
 @requires_traitsui
-class TestHTMLEditor(SimpleEditorTestMixin, unittest.TestCase):
+class TestHTMLEditor(SimpleEditorWithCachingTestMixin, unittest.TestCase):
     cache_name = "HTMLTextEditor"
     traitsui_name = "HTMLEditor"
     factory_name = "html_editor"
 
 
 @requires_traitsui
-class TestShellEditor(SimpleEditorTestMixin, unittest.TestCase):
+class TestShellEditor(SimpleEditorWithCachingTestMixin, unittest.TestCase):
     cache_name = "PythonShellEditor"
     traitsui_name = "ShellEditor"
     factory_name = "shell_editor"
 
 
 @requires_traitsui
-class TestTimeEditor(SimpleEditorTestMixin, unittest.TestCase):
+class TestTimeEditor(SimpleEditorWithCachingTestMixin, unittest.TestCase):
     cache_name = "TimeEditor"
     traitsui_name = "TimeEditor"
     factory_name = "time_editor"

